zerver/webhooks/gitlab/view.py

```diff
@@ -265,12 +265,29 @@ def get_pipeline_event_body(payload: Dict[str, Any]) -> str:
             project_homepage,
             build.get('id')
         )
-        builds_status += "* [{}]({}) - {}\n".format(
+        artifact_filename = build.get("artifacts_file").get("filename")
```

Comment:
Just a small stylistic nitpick, if we're not going to use `dict.get(key, default)` (i.e. specify a default) then it would be better to write `build.get("artifacts_file").get("filename")` as `build.get["artifacts_file"]["filename"]`. Could you change this?

Reply:
Sorry for trying to explain my point on any comment - I'm taking your comments and hints to learn :)
My intention for the double `.get()` is to be error-prone.
https://github.com/zulip/zulip/blob/20ffce52389075e3aa10affd57aea80b89983a91/zerver/webhooks/gitlab/view.py#L268-L278
should provide an empty `artifact_string` - no matter what could go wrong accessing the webhook data.

Reply:
> My intention for the double .get() is to be error-prone since the get()'s default return would be None.

The first issue with this is that when the first `get` call returns, if the key does not exist, it would return `None` just as you said. But then you'd be trying to do the second `get` against this `None` which is invalid.

```python3
>>> x = {}
>>> x.get("x")
>>> x.get("x").get("x")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'get'
```

If you really wanted this default behavior, then it's always better to specify it explicitly to let others know that this is indeed the behavior you want (and this applies to software development in general). In this case, explicitly specifying `build.get("artifacts_file", None)` would be better than leaving it at `build.get("artifacts_file")`. Though I think what you really want here is something more like `build.get("artifacts_file", {}).get("filename", None)` or something like that.

Reply:
Thank you for pointing this out 🤦  - and for taking your time for a comprehensive and extensive explanation!
I went for your ultimate suggestion
```python
build.get("artifacts_file", {}).get("filename", None)
```

Reply:
Looks good. We can resolve this thread now :slightly_smiling_face: